lib/slave/cracker.go

```diff
@@ -8,14 +8,20 @@ import (
 	"github.com/aaronang/cong-the-ripper/lib/slave/hasher"
 )
 
-func Execute(task lib.Task, successChan chan CrackerSuccess, failChan chan CrackerFail) {
-	bd := brutedict.New(&task)
+func Execute(task *task, successChan chan CrackerSuccess, failChan chan CrackerFail) {
+	bd := brutedict.New(&task.Task)
 	var hasher hasher.Hasher = new(hasher.Pbkdf2) // Can be swapped with other hashing algorithms
 
 	log.Println("[ Task", task.ID, "]", "Start cracker.Execute")
 	for {
 		if candidate := bd.Next(); candidate != nil {
-			hash := hasher.Hash(candidate, &task)
+			select {
+			case c := <-task.progressChan:
+				c <- string(candidate)
+			default:
+			}
+
+			hash := hasher.Hash(candidate, &task.Task)
```

Comment:
I don't quite understand what different it would make if we don't check next candidate first thing in `for`, does my proposal fail in some cases?

I just realised, the current implementation isn't gonna work if the slave crashes after the progress is updated but before `hasher.Hash` is completed.